climada/util/api_client.py

```diff
@@ -179,18 +251,45 @@ class AmbiguousResult(Exception):
     class NoResult(Exception):
         """Custom Exception for No Query Result"""
 
-    def __init__(self):
+    class NoConnection(Exception):
+        """To be raised if there is no internet connection and no cached result."""
+
+
+    @staticmethod
+    def _is_online(url):
+        host = [x for x in url.split('/')
+                if x not in ['https:', 'http:', '']][0]
+        port = 80 if host.startswith('http://') else 443
+        socket.setdefaulttimeout(1)
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as skt:
+            try:
+                skt.connect((host, port))
+                return True
+            except socket.error:
+                return False
+
+    def __init__(self, cache_enabled=None):
         """Constructor of Client.
 
         Data API host and chunk_size (for download) are configurable values.
         Default values are 'climada.ethz.ch' and 8096 respectively.
+
+        Parameters
+        ----------
+        cache_enabled : bool, optional
+            This flag controls whether the api calls of this client are going to be cached to the
+            local file system (location defined by CONFIG.data_api.cache_dir).
+            If set to true, the client can reload the results from the cache in case there is no
+            internet connection and thus work in offline mode.
+            Default: None, in this case the value is taken from CONFIG.data_api.cache_enabled.
         """
         self.headers = {"accept": "application/json"}
         self.url = CONFIG.data_api.url.str().rstrip("/")
         self.chunk_size = CONFIG.data_api.chunk_size.int()
+        self.cache = Cacher(cache_enabled)
+        self.online = Client._is_online(self.url)
```

Comment:
how expensive is `_is_online` to call in terms of API load? would it make sense to evaluate `Client.is_online` whenever an API call is made (and `cache_enabled = True`), rather than when the client is created?

if this is more than ten minutes to implement it's probably not worth it. i'm thinking about it because my train wifi is disconnecting and reconnecting as i write.